source/opt/folding_rules.cpp

```diff
@@ -1907,6 +1907,27 @@ FoldingRule RedundantFMix() {
   };
 }
 
+// This rule handles addition of zero for integers.
+FoldingRule RedundantIAdd() {
+  return [](IRContext*, Instruction* inst,
+            const std::vector<const analysis::Constant*>& constants) {
+    assert(inst->opcode() == SpvOpIAdd && "Wrong opcode. Should be OpIAdd.");
+
+    uint32_t operand = std::numeric_limits<uint32_t>::max();
+    if (constants[0] && constants[0]->IsZero())
+      operand = inst->GetSingleWordInOperand(1);
+    else if (constants[1] && constants[1]->IsZero())
+      operand = inst->GetSingleWordInOperand(0);
+
+    if (operand != std::numeric_limits<uint32_t>::max()) {
+      inst->SetOpcode(SpvOpCopyObject);
```

Comment:
OpCopyObject requires result type to match operand type.  But OpIAdd can be used with mixed signedness operands.
If the types are different, use an OpBitcast instead.

And then if we want to be thorough we might want to have folding rules for bitcast-feeding-X where X has flexibility on the type of its operands. But that can come later.

Reply:
Thanks, I didn't realize that.